### ISSUE TITLE
CVC4 refutation and solution soundness bugs in the QF_FP logic

### ISSUE THREAD

**Issue by muchang** [CVC4/CVC4#5504]
For this formula, CVC4 incorrectly gives sat on this unsat formula
```
[520] % z3release small.smt2
unsat
[521] % cvc4 small.smt2
sat
[522] % 
[522] % cat small.smt2
(set-logic ALL)
(define-fun a () (_ FloatingPoint 11 53) (fp #b1 #b11101110101 #b1111000011000000101000110100011000011010100000111111))
(define-fun b () (_ FloatingPoint 11 53) (fp #b0 #b11111011001 #b0111110010010011001010011101111011010100101110011010))
(define-fun c () (_ FloatingPoint 11 53) (fp #b1 #b11111111111 #b0000000000000000000000000000000000000000000000000000))
(assert (distinct (fp.mul roundTowardNegative a b) c))
(check-sat)
[523] %
```
Commit: 36af095

**Comment by muchang** [comment]
Here is another refutation soundness bug that might be related:
```
[521] % z3release small.smt2
sat
[522] % cvc4 small.smt2
unsat
[523] % 
[523] % cat small.smt2
(set-logic QF_FPBV)
(declare-fun a () (_ BitVec 32))
(assert
(let ((b ((_ to_fp 8 24) (_ bv0000000001 32))))
(let ((c (fp.add roundTowardPositive (fp.add roundTowardPositive ((_ to_fp 8 24) a) b) b)))
(let ((d (fp.add roundTowardPositive (fp.add roundTowardPositive c b) b)))
(let ((f (fp.add roundTowardNegative (fp.add roundTowardNegative ((_ to_fp 8 24) a) b) b)))
(let ((e (fp.add roundTowardNegative (fp.add roundTowardNegative f b) b)))
(xor (not (fp.isNaN d)) (fp.geq e d))))))))
(check-sat)
[524] %
```

**Comment by zhendongsu** [comment]
Another related instance:
```
[621] % z3release small.smt2
sat
[622] % cvc4 small.smt2
unsat
[623] % cat small.smt2
(set-logic QF_FP)
(assert (fp.eq ((_ to_fp 8 24) roundTowardNegative (fp #b0 #b10001111111 #x0000000000000)) (fp #b0 #xfe #b11111111111111111111111)))
(check-sat)
[624] % 
```

**Comment by aniemetz** [comment]
Fixed in #5578.
